### PR TITLE
CSVReader: reading rows iteratively, avoid recursion

### PR DESCRIPTION
**Pull request by minotogna**
Using CSVReader iteratively made us save approximately 80MB of memory during TaxonomiesImportJob and SamplingPointDataImportJob running within CollectImportJob

### CODE REVIEW

File: server/utils/file/csvReader.js [hunk 1]
```diff
@@ -7,88 +7,70 @@ import * as StringUtils from '@core/stringUtils'
...
     }
 
-    const processNext = () => {
+    const _indexRowByHeaders = row => headers
+      ? headers.reduce(
+        (accRow, header, index) => Object.assign(accRow, { [header]: StringUtils.trim(row[index]) }),
+        {}
+      )
+      : row
+
```
Comment (sgiaccio) [inline server/utils/file/csvReader.js]: Would R.indexBy be more readable? trim would have to be done separately though

File: server/utils/file/csvReader.js [hunk 2]
```diff
@@ -7,88 +7,70 @@ import * as StringUtils from '@core/stringUtils'
...
       (async () => {
-        processingRow = true
 
-        if (queue.isEmpty()) {
-          if (ended)
-            resolve()
-        } else if (!canceled) {
-          const row = queue.dequeue()
+        // run until there's a row in the queue and it's not been canceled
+        for (let row = queue.dequeue(); !!row && !canceled; row = queue.dequeue()) {
```
Comment (sgiaccio) [inline server/utils/file/csvReader.js]: while ((row = queue.dequeue()) !== null && !canceled) to avoid double assignment?

Comment (minotogna) [reply server/utils/file/csvReader.js]: ok, i changed it

File: server/utils/file/csvReader.js (hunk 1 above)
Comment (minotogna) [reply server/utils/file/csvReader.js]: R.indexBy is not suitable in this context

File: server/utils/file/csvReader.js [hunk 3]
```diff
@@ -7,88 +7,70 @@ import * as StringUtils from '@core/stringUtils'
...
-          if (ended)
-            resolve()
-        } else if (!canceled) {
-          const row = queue.dequeue()
+    const processQueue = () => {
+      (async () => {
 
+        let row
+        // run until there's a row in the queue and it's not been canceled
+        while (!!(row = queue.dequeue()) && !canceled) {
```
Comment (eliask) [inline server/utils/file/csvReader.js]: `!!` is unnecessary here, the truth-value is the same regardless.

Comment (minotogna) [reply server/utils/file/csvReader.js]: I know, i added it just to self document the code. If you prefer, I'll remove it

Comment (minotogna) [reply server/utils/file/csvReader.js]: I'll ask @SteRiccio to remove this in his next pull request

**Pull request merged**